File: backend/utils/validators.js
const Joi = require('joi');

const validators = {
  phone: Joi.string()
    .pattern(/^(\+237|237)?[2368]\d{8}$/)
    .required()
    .messages({
      'string.pattern.base': 'Numéro de téléphone invalide',
      'any.required': 'Le numéro de téléphone est requis',
    }),
  
  otp: Joi.string()
    .length(6)
    .pattern(/^\d+$/)
    .required()
    .messages({
      'string.length': 'Le code doit contenir 6 chiffres',
      'string.pattern.base': 'Le code doit contenir uniquement des chiffres',
    }),
  
  fullName: Joi.string()
    .min(2)
    .max(100)
    .required()
    .messages({
      'string.min': 'Le nom doit contenir au moins 2 caractères',
      'string.max': 'Le nom est trop long',
    }),
  
  referralCode: Joi.string()
    .pattern(/^[A-Z]{3,4}-\d{4}$/)
    .optional()
    .messages({
      'string.pattern.base': 'Format de code invalide (ex: SHIELD-1234)',
    }),
};

module.exports = { validators };

File: shieldme/lib/core/utils/validators.dart
class Validators {
  static String? validatePhone(String? value) {
    if (value == null || value.isEmpty) {
      return 'Veuillez entrer votre numéro de téléphone';
    }
    // Remove spaces and special characters for validation
    final cleanPhone = value.replaceAll(RegExp(r'[\s\-\(\)\+]'), '');
    if (cleanPhone.length < 9) {
      return 'Numéro de téléphone invalide';
    }
    return null;
  }

  static String? validateName(String? value) {
    if (value == null || value.isEmpty) {
      return 'Veuillez entrer votre nom';
    }
    if (value.length < 2) {
      return 'Nom trop court';
    }
    if (value.length > 50) {
      return 'Nom trop long';
    }
    return null;
  }

  static String? validateOTP(String? value) {
    if (value == null || value.isEmpty) {
      return 'Veuillez entrer le code OTP';
    }
    if (value.length != 6) {
      return 'Le code doit contenir 6 chiffres';
    }
    if (!RegExp(r'^\d+$').hasMatch(value)) {
      return 'Le code ne doit contenir que des chiffres';
    }
    return null;
  }

  static String? validateReferralCode(String? value) {
    if (value == null || value.isEmpty) {
      return null; // Optional field
    }
    if (!RegExp(r'^[A-Z]{3,4}-\d{4}$').hasMatch(value.toUpperCase())) {
      return 'Format invalide (ex: SHIELD-1234)';
    }
    return null;
  }

  static String? validateUrl(String? value) {
    if (value == null || value.isEmpty) {
      return 'Veuillez entrer un lien';
    }
    final urlPattern = RegExp(
      r'^(https?:\/\/)?' // http:// or https://
      r'([\da-z\.-]+)\.([a-z\.]{2,6})' // domain
      r'([\/\w \.-]*)*\/?$', // path
      caseSensitive: false,
    );
    if (!urlPattern.hasMatch(value)) {
      return 'Lien invalide';
    }
    return null;
  }
}